### ISSUE TITLE
Reliability of Prometheus exposed metrics names

### ISSUE THREAD

**Issue by jdel** [netdata/netdata#5527]
##### Question summary
Prometheus metrics names have changed between v1.11.1 and v1.12.1. I have not tried v1.12.0.
I am fairly new to Netdata, started with v1.11.1 and upgraded to v1.12.1 yesterday to notice the name of many metrics exposed to prometheus on `/api/v1/allmetrics?format=prometheus` have changed, effectively wrecking everything depending on prometheus in the background.

Being in the early stages of using Netdata as a scraping endpoint for Prometheus, i can work around those changes, but the main questions are:
  - How much can I rely on these metrics names in the future ?
  - Was this change announced prior to implementation or described in a changelog (I couldn't find any) ?

Diff after stripping all labels. The list is non exhaustive.

<details>
Left is v1.11.1, right is v1.12.1:

```diff
1,3c1,3
< netdata_apps_cpu_cpu_time___average
< netdata_apps_cpu_system_cpu_time___average
< netdata_apps_cpu_user_cpu_time___average
---
> netdata_apps_cpu_percentage_average
> netdata_apps_cpu_system_percentage_average
> netdata_apps_cpu_user_percentage_average
5,6c5,6
< netdata_apps_lreads_kilobytes_persec_average
< netdata_apps_lwrites_kilobytes_persec_average
---
> netdata_apps_lreads_KiB_persec_average
> netdata_apps_lwrites_KiB_persec_average
8c8
< netdata_apps_mem_MB_average
---
> netdata_apps_mem_MiB_average
11c11
< netdata_apps_preads_kilobytes_persec_average
---
> netdata_apps_preads_KiB_persec_average
13c13
< netdata_apps_pwrites_kilobytes_persec_average
---
> netdata_apps_pwrites_KiB_persec_average
15c15
< netdata_apps_swap_MB_average
---
> netdata_apps_swap_MiB_average
17,21c17,22
< netdata_apps_vmem_MB_average
< netdata_cgroup_cpu_per_core_percent_average
< netdata_cgroup_cpu_percent_average
< netdata_cgroup_mem_MB_average
< netdata_cgroup_mem_activity_MB_persec_average
---
> netdata_apps_vmem_MiB_average
> netdata_cgroup_cpu_limit_percentage_average
> netdata_cgroup_cpu_per_core_percentage_average
> netdata_cgroup_cpu_percentage_average
> netdata_cgroup_mem_MiB_average
> netdata_cgroup_mem_activity_MiB_persec_average
23,25c24,27
< netdata_cgroup_mem_usage_MB_average
< netdata_cgroup_pgfaults_MB_persec_average
< netdata_cgroup_throttle_io_KB_persec_average
---
> netdata_cgroup_mem_usage_MiB_average
> netdata_cgroup_mem_usage_limit_MiB_average
> netdata_cgroup_pgfaults_MiB_persec_average
> netdata_cgroup_throttle_io_KiB_persec_average
27c29
< netdata_cgroup_writeback_MB_average
---
> netdata_cgroup_writeback_MiB_average
32,36c34,39
< netdata_disk_avgsz_kilobytes_per_operation_average
< netdata_disk_await_ms_per_operation_average
< netdata_disk_backlog_backlog__ms__average
< netdata_disk_inodes_Inodes_average
< netdata_disk_io_kilobytes_persec_average
---
> netdata_cpuidle_cpuidle_percentage_average
> netdata_disk_avgsz_KiB_operation_average
> netdata_disk_await_milliseconds_operation_average
> netdata_disk_backlog_milliseconds_average
> netdata_disk_inodes_inodes_average
> netdata_disk_io_KiB_persec_average
40,41c43,45
< netdata_disk_space_GB_average
< netdata_disk_svctm_ms_per_operation_average
---
> netdata_disk_qops_operations_average
> netdata_disk_space_GiB_average
> netdata_disk_svctm_milliseconds_operation_average
43,45c47,49
< netdata_groups_cpu_cpu_time___average
< netdata_groups_cpu_system_cpu_time___average
< netdata_groups_cpu_user_cpu_time___average
---
> netdata_groups_cpu_percentage_average
> netdata_groups_cpu_system_percentage_average
> netdata_groups_cpu_user_percentage_average
47,48c51,52
< netdata_groups_lreads_kilobytes_persec_average
< netdata_groups_lwrites_kilobytes_persec_average
---
> netdata_groups_lreads_KiB_persec_average
> netdata_groups_lwrites_KiB_persec_average
50c54
< netdata_groups_mem_MB_average
---
> netdata_groups_mem_MiB_average
53c57
< netdata_groups_preads_kilobytes_persec_average
---
> netdata_groups_preads_KiB_persec_average
55c59
< netdata_groups_pwrites_kilobytes_persec_average
---
> netdata_groups_pwrites_KiB_persec_average
57c61
< netdata_groups_swap_MB_average
---
> netdata_groups_swap_MiB_average
59c63
< netdata_groups_vmem_MB_average
---
> netdata_groups_vmem_MiB_average
62,64d65
< netdata_ip_inerrors_packets_persec_average
< netdata_ip_mcast_kilobits_persec_average
< netdata_ip_mcastpkts_packets_persec_average
66a68,71
> netdata_ipv4_errors_packets_persec_average
> netdata_ipv4_icmp_errors_packets_persec_average
> netdata_ipv4_icmp_packets_persec_average
> netdata_ipv4_icmpmsg_packets_persec_average
69c74
< netdata_ipv4_sockstat_tcp_mem_KB_average
---
> netdata_ipv4_sockstat_tcp_mem_KiB_average
71c76
< netdata_ipv4_sockstat_udp_mem_KB_average
---
> netdata_ipv4_sockstat_udp_mem_KiB_average
77a83
> netdata_ipv4_udperrors_events_persec_average
81a88
> netdata_ipv6_icmperrors_errors_persec_average
89d95
< netdata_ipv6_sockstat6_raw_sockets_sockets_average
91a98
> netdata_ipv6_udperrors_events_persec_average
93,102c100,106
< netdata_ipvs_net_kilobits_persec_average
< netdata_ipvs_packets_packets_persec_average
< netdata_ipvs_sockets_connections_persec_average
< netdata_mem_available_MB_average
< netdata_mem_committed_MB_average
< netdata_mem_kernel_MB_average
< netdata_mem_pgfaults_page_faults_persec_average
< netdata_mem_slab_MB_average
< netdata_mem_transparent_hugepages_MB_average
< netdata_mem_writeback_MB_average
---
> netdata_mem_available_MiB_average
> netdata_mem_committed_MiB_average
> netdata_mem_kernel_MiB_average
> netdata_mem_pgfaults_faults_persec_average
> netdata_mem_slab_MiB_average
> netdata_mem_transparent_hugepages_MiB_average
> netdata_mem_writeback_MiB_average
118,119d121
< netdata_netdata_plugin_tc_cpu_milliseconds_persec_average
< netdata_netdata_plugin_tc_time_milliseconds_run_average
121a124,126
> netdata_netdata_registry_entries_entries_average
> netdata_netdata_registry_mem_KiB_average
> netdata_netdata_registry_sessions_sessions_average
123c128
< netdata_netdata_response_time_ms_request_average
---
> netdata_netdata_response_time_milliseconds_request_average
132c137
< netdata_netdata_tcp_connected_connected_average
---
> netdata_netdata_tcp_connected_sockets_average
135,139d139
< netdata_netfilter_conntrack_changes_changes_persec_average
< netdata_netfilter_conntrack_errors_events_persec_average
< netdata_netfilter_conntrack_expect_expectations_persec_average
< netdata_netfilter_conntrack_new_connections_persec_average
< netdata_netfilter_conntrack_search_searches_persec_average
141,142c141,143
< netdata_services_cpu_percent_average
< netdata_services_mem_usage_MB_average
---
> netdata_nfs_rpc_calls_persec_average
> netdata_services_cpu_percentage_average
> netdata_services_mem_usage_MiB_average
145,146c146,147
< netdata_services_throttle_io_read_KB_persec_average
< netdata_services_throttle_io_write_KB_persec_average
---
> netdata_services_throttle_io_read_KiB_persec_average
> netdata_services_throttle_io_write_KiB_persec_average
155c156
< netdata_system_io_kilobytes_persec_average
---
> netdata_system_io_KiB_persec_average
162c163
< netdata_system_pgpgio_kilobytes_persec_average
---
> netdata_system_pgpgio_KiB_persec_average
164c165
< netdata_system_ram_MB_average
---
> netdata_system_ram_MiB_average
168,170c169,171
< netdata_users_cpu_cpu_time___average
< netdata_users_cpu_system_cpu_time___average
< netdata_users_cpu_user_cpu_time___average
---
> netdata_users_cpu_percentage_average
> netdata_users_cpu_system_percentage_average
> netdata_users_cpu_user_percentage_average
172,173c173,174
< netdata_users_lreads_kilobytes_persec_average
< netdata_users_lwrites_kilobytes_persec_average
---
> netdata_users_lreads_KiB_persec_average
> netdata_users_lwrites_KiB_persec_average
175c176
< netdata_users_mem_MB_average
---
> netdata_users_mem_MiB_average
178c179
< netdata_users_preads_kilobytes_persec_average
---
> netdata_users_preads_KiB_persec_average
180c181
< netdata_users_pwrites_kilobytes_persec_average
---
> netdata_users_pwrites_KiB_persec_average
182c183
< netdata_users_swap_MB_average
---
> netdata_users_swap_MiB_average
184c185
< netdata_users_vmem_MB_average
---
> netdata_users_vmem_MiB_average
```
</details>

##### OS / Environment
CoreOS stable, netdata installed in `/opt/netdata` with corresponding `gz.run` from releases.

##### Component Name
`/api/v1/allmetrics?format=prometheus`

##### Expected results
I expect the metrics to not change between two minor versions of Netdata.
I expect these changes to be announced in the changelog of a preview/rc release.

Thanks for your time.

**Comment by glennie** [comment]
Hello,
I've noticed the same issue on v1.12.1, but didn't reported it.
I've lost also my metrics (charted by grafana based on promethus scrapping). If you could keep the metrics labels stable between releases and warn about changes, I will really appreciate that.
Many thanks and kind regards

**Comment by cakrit** [comment]
Gents, this was definitely not intentional. Most of the changes here have to do with the standardization of the units we did, but we definitely did not consider any possible impacts to the exporters. We'll need to think a little bit about the possible solutions we can provide as soon as possible and how to prevent any such changes in the future.

**Comment by cakrit** [comment]
Ok, so this is what we can do to resolve the main issue of the units changing. We'll add an option to the API, e.g. `legacy_unit_format=1`. This will convert the MiB to MB, KiB_persec to kilobytes_persec etc. 
It's not 100% accurate because the units have actually been standardized, but it would solve most of your problems. 

The remaining changes will be very few from the examples you provided. 
I'll start working on this, please let me know if you have any objections or better ideas.

**Comment by jdel** [comment]
As mentioned, I am too early in the Netdata adoption stage to be upset about it or require immediate mitigation, but some people might be having more trouble upgrading their Prometheus alerts or Grafana dashboards.

I am happy with any measure to prevent this happening in the future.

**Comment by paulfantom** [comment]
I think easiest one would be to create a following test:
1. create a fixture/mock and put it into `tests/`
2. start netdata in CI
3. download prometheus metrics
4. compare what is downloaded with a fixture

if anything was changed, then fail that test.

**Comment by cakrit** [comment]
> As mentioned, I am too early in the Netdata adoption stage to be upset about it or require immediate mitigation, but some people might be having more trouble upgrading their Prometheus alerts or Grafana dashboards.

I expect that most people  scrape netdata using `source=as-collected` (identical to `source=raw`), which does NOT include the units. e.g. http://localhost:19999//api/v1/allmetrics?format=prometheus&source=as-collected

It will take only about half a day to provide a tested option to show metrics with the old unit format and we have no way of knowing how many people are affected, so I'd rather do it to be on the safe side. 

The test framework will take about as long, we have a few fixtures to create and test. I'll get on this tomorrow.

**Comment by glennie** [comment]
Hi,
Many thanks for the proposed fixes and tests.
Kind regards,

**Comment by cakrit** [comment]
@jdel  can you please do another diff like the one in the OP using the code in #5331? 
It will be easier for me to check if I've missed anything. Note the caveats mentioned in the PR. 
You'll just need to add `&oldunits=yes` to the API call to get the old names.

Continuing with the other tasks...

**Comment by jdel** [comment]
Do you mean #5531 ? Is there a static build I can find somewhere ? Or instructions on how to build a specific commit ?

**Comment by cakrit** [comment]
Sorry, right. 
I do the following:

Only once:
```
wget "https://github.com/github/hub/releases/download/v2.5.1/hub-linux-amd64-2.5.1.tgz" -O "/tmp/hub-linux-amd64-2.5.1.tgz"
tar -C ~/bin -xvf "/tmp/hub-linux-amd64-2.5.1.tgz" &>/dev/null
export PATH=$PATH:"~/bin/hub-linux-amd64-2.5.1/bin"
```

And then for every PR I want to test in my git clone:

```
hub pr checkout 5531
systemctl start docker
export DEVEL=true
sudo ./packaging/docker/build.sh

sudo docker run -d --name=netdata -p 19999:19999 -v /proc:/host/proc:ro -v /sys:/host/sys:ro -v /var/run/docker.sock:/var/run/docker.sock:ro --cap-add SYS_PTRACE --security-opt apparmor=unconfined netdata:latest-amd64
```

There are [other ways to clone a specific branch](https://stackoverflow.com/questions/1911109/how-to-clone-a-specific-git-branch) as well and then you can run `netdata-installer.sh` from the root directory of the clone.

**Comment by cakrit** [comment]
If it's too much trouble, don't worry about it. I'll get the reviews, merge 5531 and then do another PR if we need more changes.

**Comment by jdel** [comment]
Hi sorry, unfortunately, the build script does not run on my current workstation (Mac, with an old bash 3.x) and I don't have time to setup a proper linux build environment.

**Comment by cakrit** [comment]
It's ok, we will merge it then.
